packages/cicero-core/src/contractinstance.js

```diff
@@ -0,0 +1,230 @@
+        const p12Asn1 = forge.asn1.fromDer(p12Der);
+        // decrypt p12 using the passphrase 'password'
+        const p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, false, passphrase);
+        //X509 cert forge type
+        const certificateForge = p12.safeContents[0].safeBags[0].cert;
+        //Private Key forge type
+        const privateKeyForge = p12.safeContents[1].safeBags[0].key;
+        //convert cert and private key from forge to PEM
+        const certificatePem = forge.pki.certificateToPem(certificateForge);
+        const privateKeyPem = forge.pki.privateKeyToPem(privateKeyForge);
+        //convert private key in pem to private key type in node
+        const privateKey = crypto.createPrivateKey(privateKeyPem);
+        const sign = crypto.createSign('SHA256');
+        sign.write(instanceHash + timestamp);
+        sign.end();
+        const signature = sign.sign(privateKey, 'hex');
+        const signatureObject = {
+            signatory,
+            instanceHash,
+            timestamp,
+            signatoryCert: certificatePem,
+            signature
+        };
+        this.contractSignatures.push(signatureObject);
+    }
+
+    /**
+     * Gets a content based SHA-256 hash for this contract instance. Hash
+     * is based on the metadata for the contract instance plus the contents of
+     * all the models and all the script files.
+     * @return {string} the SHA-256 hash in hex format
+     */
+    getHash() {
+        const content = {};
+        content.metadata = this.metadata;
+        if(this.parserManager.getTemplate()) {
+            content.grammar = this.parserManager.getTemplate();
+        }
+        content.models = {};
+        content.scripts = {};
+
+        let modelFiles = this.getModelManager().getModels();
+        modelFiles.forEach(function (file) {
+            content.models[file.namespace] = file.content;
+        });
+
+        let scriptManager = this.getScriptManager();
+        let scriptFiles = scriptManager.getScripts();
+        scriptFiles.forEach(function (file) {
+            content.scripts[file.getIdentifier()] = file.contents;
+        });
+
+        content.data = this.getData();
+
+        const hasher = crypto.createHash('sha256');
+        hasher.update(stringify(content));
+        return hasher.digest('hex');
+    }
+
+    /**
+     * Verify the signatures of the parties/individuals who signed the contract
+     */
+    verifySignatures() {
+        if(this.contractSignatures.length > 0){
+            this.contractSignatures.forEach((contractSignature)=>{
+                const { timestamp, signatoryCert, signature } = contractSignature;
+                this.verify(signature, timestamp, signatoryCert);
+            });
+        } else {
+            throw new Error('The contract is not signed by any party/individual.');
+        }
+    }
+
+    /**
+     * Verify the signatures of a partcular party/individual
+     * @param {string} signature  - signature of the signatory
```

Comment:
We need a better comment here. What is this string?

Reply:
@sanketshevkar to update.